CASSANDRA-19477 trunk dont go to disk for hints sizes

smiklosovic opened the pull request:
Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

clohfink commented:
+1 thanks for this!

smiklosovic commented on src/java/org/apache/cassandra/hints/HintsStore.java:
```diff
@@ -76,9 +77,13 @@ private HintsStore(UUID hostId, File hintsDirectory, ImmutableMap<String, Object
         //noinspection resource
         lastUsedTimestamp = descriptors.stream().mapToLong(d -> d.timestamp).max().orElse(0L);
+
+        for (HintsDescriptor descriptor : Iterables.concat(dispatchDequeue, corruptedFiles))
```
it is not technically necessary to iterate over corrputedFiles because they are empty here

iamaleksey commented on src/java/org/apache/cassandra/hints/HintsStore.java:
```diff
@@ -61,6 +61,7 @@ final class HintsStore
     private final Deque<HintsDescriptor> dispatchDequeue;
     private final Queue<HintsDescriptor> corruptedFiles;
     private final Map<HintsDescriptor, Long> hintsExpirations;
+    private final Map<String, Long> hintsSizes;
```
Why not `HintsDescriptor` as key?

smiklosovic replied:
It was like that before, then I changed it to String. Is not this just better? I don't mind to revert this.

iamaleksey commented on src/java/org/apache/cassandra/hints/HintsWriteExecutor.java:
```diff
@@ -260,19 +261,33 @@ private void flushInternal(Iterator<ByteBuffer> iterator, HintsStore store)
-                if (session.position() >= maxHintsFileSize)
-                    break;
+                Map<String, Long> hintsSizes = store.getHintsSizes();
+                hintsSizes.compute(descriptor.hintsFileName, (d, currentSize) ->
```
You have `session.position()` ~~and `computeIfAbsent()`~~.

iamaleksey replied:
But really just `put()` is sufficient.

iamaleksey replied:
(This is a capturing lambda fyi)

iamaleksey commented on src/java/org/apache/cassandra/hints/HintsStore.java:
```diff
@@ -282,7 +297,11 @@ boolean isWriting()
         if (hintsWriter == null)
+        {
             hintsWriter = openWriter();
+            hintsSizes.put(hintsWriter.descriptor().fileName(), hintsWriter.descriptor().file(hintsDirectory).length());
```
This seems pretty unnecessary or at the very least unjustified.

smiklosovic replied:
we do not include .crc32 files too btw ... I can remove this though. There is a test at the bottom which thinks that by mere writing of a descriptor it should has some size. I can rewrite that test accordingly.

iamaleksey replied:
CRC files don't matter, an entirely different order of magnitude where total limits are involved.

iamaleksey commented on src/java/org/apache/cassandra/hints/HintsStore.java:
```diff
@@ -203,14 +213,21 @@ private void deleteHints(Predicate<HintsDescriptor> predicate)
         File hintsFile = descriptor.file(hintsDirectory);
         if (hintsFile.tryDelete())
+        {
+            // this method is also called standalone (without deleteHints)
```
You are calling this twice on the normal path. Just leave the one call in this method, in `finally` if need be.

iamaleksey commented:
I would just cache the size value in a transient field in `HintsDescriptor`. A CHM is an overkill here.

iamaleksey commented on src/java/org/apache/cassandra/hints/HintsStore.java:
```diff
@@ -282,7 +293,11 @@ boolean isWriting()
         if (hintsWriter == null)
+        {
             hintsWriter = openWriter();
+            hintsSizes.put(hintsWriter.descriptor(), hintsWriter.descriptor().file(hintsDirectory).length());
```
I don't think we need to do this.

iamaleksey commented on src/java/org/apache/cassandra/hints/HintsWriteExecutor.java:
```diff
@@ -260,19 +261,30 @@ private void flushInternal(Iterator<ByteBuffer> iterator, HintsStore store)
-                session.append(iterator.next());
-                if (session.position() >= maxHintsFileSize)
-                    break;
+                store.getHintsSizes().put(descriptor, session.position());
```
No need to put this in a finally block I'd say.

smiklosovic committed:
Do not go to disk for reading hints file sizes

patch by Stefan Miklosovic; reviewed by Aleksey Yeschenko, Chris Lohfink for CASSANDRA-19477

smiklosovic commented on src/java/org/apache/cassandra/hints/HintsWriter.java:
```diff
@@ -259,7 +261,7 @@ void append(Hint hint) throws IOException
             if (hintBuffer == buffer)
                 bytesWritten += totalSize;
             else
-                append((ByteBuffer) hintBuffer.flip());
```
not necessary

iamaleksey replied:
Not anymore in Java 11, was needed in Java 8 :)

smiklosovic commented on src/java/org/apache/cassandra/hints/HintsWriter.java:
```diff
@@ -224,7 +226,7 @@ void append(ByteBuffer hint) throws IOException
          * Serializes and appends the hint (with CRC included) to this session's aggregation buffer,
          * writes to the underlying channel when the buffer is overflown.
          *
-         * Used mainly by tests and {@link LegacyHintsMigrator}
```
does not exist anymore

smiklosovic commented on src/java/org/apache/cassandra/hints/HintsWriter.java:
```diff
@@ -271,6 +273,7 @@ public void close() throws IOException
             flushBuffer();
             maybeFsync();
             maybeSkipCache();
+            descriptor.size = position();
```
why not to just update it in close() ? Then when this is used in try-with-resources in HintsWriteExecutor, we do not need to change there anything at all and not mess with finally etc ...

smiklosovic replied:
this should be probably in finally though, not sure ... as all above can throw

iamaleksey commented on src/java/org/apache/cassandra/hints/HintsDescriptor.java:
```diff
@@ -81,18 +81,26 @@ final class HintsDescriptor
 
+    // size of underlying hint file (without descriptor header and crc file sizes)
+    // the value is populated in Session's close method in HintsWriter
+    volatile long size = 0;
```
`size` -> `fileSize` slightly more descriptive. May also mark it as `transient`, to borrow the Accord convention (for the reader, not for the JVM).

smiklosovic replied:
why not
